### ISSUE TITLE
Websocket provider subscribe_logs hangs

### ISSUE THREAD

**Issue by mlagoma** [gakonst/ethers-rs#1620]
**Version**
0.17.0

**Platform**
Linux archdom 5.18.15-arch1-1 x86_64 GNU/Linux

**Description**
When running the example code for subscribe_logs with a websocket from Hardhat, the program gets stuck in the while loop after 5-10 minutes and no longer picks up new events. The program must be killed and restarted to continue picking up events, including the ones that were not picked up.

To reproduce the issue:
1. Start a hardhat network node and run a contract function that produces an event.
2. Run the [example code for subscribe_logs](https://docs.rs/crate/ethers/0.17.0/source/examples/subscribe_logs.rs) with the websocket from the hardhat node.
3. Continue triggering events on the hardhat node, particularly after waiting 5-10 minutes.

I've created a github repo to help setup this environment:
https://github.com/mlagoma/ethers-subscribe-logs-debug

Ideally the while loop would continue to pick up events indefinitely.
It might also be nice to have some way to check if a "stream.next()" exists before awaiting. I tried to find a way to do this with current methods available but I was not able to find anything much more than stream.next(). This could be one way to workaround the issue, perhaps useful for other use cases.

**Comment by mattsse** [comment]
Thanks os much for the repro!

I was able to reproduce this with hardhat,

I tried to reproduce this with anvil and fixed a couple of bugs there:
https://github.com/foundry-rs/foundry/pull/2856
https://github.com/foundry-rs/foundry/pull/2852

with these anvil fixes I wasn't able to reproduce this anymore,

for hardhat, it looks like the connection is not closed but there are no new messages after a while (like ~5min).

could you try to replicate this with [anvil](https://github.com/foundry-rs/foundry/tree/master/anvil) on master?

`anvil -p 8888` is equivalent. there is an issue with the network identifier though, not familiar enough with hardhat's providers to recommend a fix, but I worked around this by commenting out the check that causes `OnlyHardhatNetworkError: This helper can only be used with Hardhat Network. You are connected to 'development', whose identifier is 'anvil/v0.1.0'`

**Comment by mlagoma** [comment]
Thank you so much for the quick response!

The `OnlyHardhatNetworkError` must have something to do with this test contract I took from the hardhat getting started page. I've tried out the real contracts I care about on anvil and indeed it is working correctly, without this problem.

I've found an issue in the hardhat repo (https://github.com/NomicFoundation/hardhat/issues/2053) that confirms your suspicion that this is an issue with hardhat.

I think I can go ahead and continue testing my subscribe_logs code using our contracts deployed to anvil, so I will go ahead and close the issue.

Note: I do encounter an error on anvil that I don't encounter on hardhat / testnets with the real contracts: `Error: reverted without a reason string`. This doesn't cause any issues with subscribe_logs though, so I think anvil will work good enough. If necessary I can find a way to workaround hardhat's issue.
